client/X11/xf_rail.c

```diff
@@ -269,90 +303,153 @@ void xf_rail_paint(xfContext* xfc, INT32 uleft, INT32 utop, UINT32 uright,
 
 /* RemoteApp Core Protocol Extension */
 
-static BOOL xf_rail_window_common(rdpContext* context,
-                                  const WINDOW_ORDER_INFO* orderInfo, const WINDOW_STATE_ORDER* windowState)
+static BOOL resizeMarginXSet = FALSE;
+static UINT32 resizeMarginLeft = 0;
+static UINT32 resizeMarginRight = 0;
+
+static BOOL resizeMarginYSet = FALSE;
+static UINT32 resizeMarginTop= 0;
+static UINT32 resizeMarginBottom = 0;
+
+static xfAppWindow* xf_rail_window_new(xfContext* xfc,
+                                       const WINDOW_ORDER_INFO* orderInfo,
+                                       const WINDOW_STATE_ORDER* windowState)
 {
-	xfAppWindow* appWindow = NULL;
-	xfContext* xfc = (xfContext*) context;
 	UINT32 fieldFlags = orderInfo->fieldFlags;
-	BOOL position_or_size_updated = FALSE;
-	appWindow = xf_rail_get_window(xfc, orderInfo->windowId);
+	xfAppWindow* appWindow = xf_rail_get_window(xfc, orderInfo->windowId);
 
-	if (fieldFlags & WINDOW_ORDER_STATE_NEW)
+	if (!appWindow)
 	{
-		if (!appWindow)
-			appWindow = xf_rail_add_window(xfc, orderInfo->windowId, windowState->windowOffsetX,
-			                               windowState->windowOffsetY, windowState->windowWidth,
-			                               windowState->windowHeight, 0xFFFFFFFF);
+		appWindow = xf_rail_add_window(xfc, orderInfo->windowId, windowState->windowOffsetX,
+		                               windowState->windowOffsetY, windowState->windowWidth,
+		                               windowState->windowHeight, 0xFFFFFFFF);
+	}
 
-		if (!appWindow)
-			return FALSE;
+	if (!appWindow)
+		return NULL;
 
-		appWindow->dwStyle = windowState->style;
-		appWindow->dwExStyle = windowState->extendedStyle;
+    appWindow->dwStyle = windowState->style;
+    appWindow->dwExStyle = windowState->extendedStyle;
+
+    if ((orderInfo->fieldFlags & WINDOW_ORDER_FIELD_RESIZE_MARGIN_X) == 0 && resizeMarginXSet) {
```

Comment:
Kind of useless, `appWindow` is created using `calloc` which initialized everything with `0`

Reply:
I don't understand which part is useless. Those lines (331-334) don't use any values from appWindow. `windowState->style` and `extendedStyle` aren't always 0, are they? (Since they come from the server I didn't think we could assume that, anyway.)

Reply:
The static variables above are kind of useless.
You get the resize margins from the window state (an argument to this function)
If the flag is set, use it, otherwise just ignore the field

Reply:
Actually, now that I think about it some more, I don't think line 334 is needed at all. If the new window has its own margins, those will get set later (starting at 409) and override the ones set here, anyway. I'll clean it up.

In case it's not clear, the idea with `resizeMarginXSet` is to enable remembering the first margins that are encountered so they can be used as default margins for later windows that don't have any specified. Another approach would be to always use the most recent margins as the default for new windows that don't have them. I don't know which is best.

Reply:
I wouldn´t do any client side magic as there are many types of windows that have no margins on purpose (think menus and tooltips, ...)

Reply:
I would agree, except that unfortunately, things break without any client side magic. For example, if margins are only used for windows that report them, then the Font dialog box for the notepad application doesn't work right. It has the moving and resizing behavior that started this issue. If the client remembers the margin from the previous window that had one and reuses it for the Font dialog box, it works. I've seen the same behavior with a dialog box for view options in explorer.exe. It appears that the server expects margins to exist on dialog boxes but does not specify them in its messages to the client.

I understand your concern about possible effects of this on windows that perhaps should not have margins. Experimentally, though, there does not seem to be any ill effect on menus or tooltips.

Reply:
you might want to take a look at how `ClientWindowMove` is used.
I don´t have the time currently as I have work to be done on some other projects, but what I can tell is that the server is sending the correct margins and if reported correctly does not start to resize the window.
Might need some delay magic as the `ClientWindowMove` should only be used if the window is moved without the server noticing (by window manager or similar) and if both, the server try to move and the client reports back too often there might be a mixup.

Reply:
I will look at `ClientWindowMove` and investigate whether delay magic might help. I have some questions about some of your comments:

1. "the server is sending the correct margins". How can I find out what the "correct" margins are?
1. "if reported correctly does not start to resize the window". Could you elaborate? If what is "reported correctly"? When using the nightly build, any time I move a window with the window manager or by grabbing the title bar, it shrinks.

Reply:
1. The margins that are announced for a window ID are the correct ones. No magic necessary (just like written in the protocol spec, default is no margin)
1. This is the exact issue we´re talking about here. 
  * During normal operation the server is placing the window where it should be (you resize with mouse or drag the title bar)
  * When dragging via window manager the server does not know where the window is and for this situation `ClientWindowMove` is responsible
  * `ClientWindowMove` needs to announce the window size with the margins
  * The server announces the winow size without margins (the margins must be manually added using the values previously announced)
So, the issue is somewhere one (or both) of two things happen:
1. The window size is announced or applied with the wrong margins
1. The `ClientWindowMove` is sending new window geometry while the server is still trying to apply other settings (network lag or similar)

Reply:
Here's what I observed with dialog boxes. This example uses the Font dialog box from notepad.exe, but I've also seen it with a view option dialog box from explorer.exe.

I'll attach the full output log, but for clarity I have only included what I think are the relevant bits below.

I started up with `xfreerdp /v:xxx /u:xxx /p:xxx /app:'c:\windows\notepad.exe' /span /log-filters:com.freerdp.client.x11:debug,com.freerdp.channels.rail.client:debug,com.freerdp.core.update:debug`.

The main window showed up with this output:
```
[12:20:01:166] [2272:2285] [DEBUG][com.freerdp.core.update] - WindowCreate windowId=0x131332 owner=0x0 [ex]style=<0x14cf0000, 0x110 visible thickframe border caption> show=current title clientOffsetX=851 clientOffsetY=825 resizeMarginLeft=7 resizeMarginRight=7 resizeMarginTop=0 resizeMarginBottom=7 windowOffsetX=425 windowOffsetY=387 windowClientDeltaX=426 windowClientDeltaY=438 windowWidth=842 windowHeight=351 windowRects=((0,0,842,351)) visibleOffsetX=425 visibleOffsetY=387 visibilityRects=((0,0,842,351))
```
Notice that margins are set.

I brought up the Font dialog box and got this output:
```
[12:32:37:102] [2272:2285] [DEBUG][com.freerdp.core.update] - WindowCreate windowId=0x655614 owner=0x20104 [ex]style=<0x94c800c4, 0x10101 popup visible border caption> show=current title clientOffsetX=835 clientOffsetY=1065 windowOffsetX=417 windowOffsetY=517 windowClientDeltaX=418 windowClientDeltaY=548 windowWidth=427 windowHeight=471 windowRects=((0,0,427,471)) visibleOffsetX=417 visibleOffsetY=517 visibilityRects=((0,0,427,471))
```
Notice that **no** margins are set.

I used `wmctrl` to move the dialog box window to (0,0):
```
[12:34:04:127] [2272:2285] [DEBUG][com.freerdp.client.x11] - updated internal geometry from X: 427x471+0+0
```

The window moved instantly but its size did not change, and no further output was generated until the window got focus, and then:
```
[12:35:40:857] [2272:2285] [DEBUG][com.freerdp.client.x11] - forwarding geometry from X/internal to RDP server: 427x471+0+0	(0,0), (427,471) {this is the ClientWindowMove}
[12:35:40:878] [2272:2285] [DEBUG][com.freerdp.core.update] - WindowUpdate windowId=0x655614 clientOffsetX=15 clientOffsetY=31 windowOffsetX=7 windowOffsetY=0 windowClientDeltaX=8 windowClientDeltaY=31 windowWidth=413 windowHeight=464 windowRects=((0,0,413,464)) visibleOffsetX=7 visibleOffsetY=0 visibilityRects=((0,0,413,464))
[12:35:40:878] [2272:2285] [DEBUG][com.freerdp.client.x11] - setting new internal & X geometry: 413x464+7+0	showState=5
[12:35:40:894] [2272:2285] [DEBUG][com.freerdp.client.x11] - updated internal geometry from X: 413x464+7+0
```
The client did not adjust the position or size of the window in the ClientWindowMove, because no margins were specified by the server for this window.

After the ClientWindowMove, the server immediately responded with a WindowUpdate that had a location and size that were different from the ClientWindowMove, even though no margin has been specified for this window. The amount of the difference is exactly the margins specified for the main window.

As shown in the log, over 30 seconds elapsed between operations, so I don't see a way to improve the situation by adding delay to the client somewhere.

I think the server is using the same margins for the dialog box as for the main window even though it does not send those margins to the client when the dialog box is created.

I am open to other explanations and/or suggested fixes.

Reply:
I tried to attach the full log but it's not working. I'll try again later. (Tried with Firefox and Chrome, both drag & drop and dialog box.)

Reply:
hmm, you can only attach zip and image files ;)
As for the resizing, which os version are you connecting to?
the resize_margin flag is only supported with newer windows versions if I read the spec correctly.
might need to take the client area and offsets into account too.

Reply:
> hmm, you can only attach zip and image files ;)

I zipped it and tried that, but still doesn't work. Do I need some kind of permission to attach files to this issue tracker? I couldn't find anything helpful in the github help.

> As for the resizing, which os version are you connecting to?
> the resize_margin flag is only supported with newer windows versions if I read the spec correctly.

Windows Server 2016 Datacenter. Also, the server does send window margins for the main window.

> might need to take the client area and offsets into account too.

Hmmm. Sure, could be. Looking at the values the server reports for those, I don't understand what they mean or how they could help. The RDP spec is not very informative. Do you know of any documentation with more info?